### ISSUE TITLE
[BUG] Unexpectedly high memory consumption by Lucene

### ISSUE THREAD

**Issue by PavelArzhaev** [opensearch-project/k-NN#507]
Hi!
I’m trying to upload 10s of million 512-dim vectors using the OpenSearch KNN plugin.
The problem is disk storage gets filled too quickly.
I do not specify the ANN index in props:

```
"properties": {
    "field_1": {
        "type": "keyword",
        "index": False,
        "norms": False,
        "ignore_above": 36
    },
    "field_2": {
        "type": "keyword",
        "norms": False,
        "ignore_above": 36
    },
    "vector_field": {
        "type": "knn_vector",
        "dimension": 512
    },
}
```

Using the calculation (1.1 * (4 * dimension + 8 * M), in my case dimension=512, M=16), 1 million vectors should require ~2.4Gb, but, as I mentioned, in my test it’s 10+Gb.

Those are stats for my test instance, 2573511 docs uploaded, ~29Gb taken, which means every doc is ~12Kb:
```
"primaries": {
	"docs": {
		"count": 2573511,
		"deleted": 0
	},
	"store": {
		"size_in_bytes": 31388403696,
		"reserved_in_bytes": 0
	},
	"indexing": {
		"index_total": 2573511,
		"index_time_in_millis": 1477518,
		"index_current": 0,
		"index_failed": 0,
		"delete_total": 0,
		"delete_time_in_millis": 0,
		"delete_current": 0,
		"noop_update_total": 0,
		"is_throttled": false,
		"throttle_time_in_millis": 0
	},
	"get": {
		"total": 0,
		"time_in_millis": 0,
		"exists_total": 0,
		"exists_time_in_millis": 0,
		"missing_total": 0,
		"missing_time_in_millis": 0,
		"current": 0
	},
	"search": {
		"open_contexts": 0,
		"query_total": 54,
		"query_time_in_millis": 55,
		"query_current": 0,
		"fetch_total": 21,
		"fetch_time_in_millis": 14,
		"fetch_current": 0,
		"scroll_total": 0,
		"scroll_time_in_millis": 0,
		"scroll_current": 0,
		"suggest_total": 0,
		"suggest_time_in_millis": 0,
		"suggest_current": 0
	},
	"merges": {
		"current": 0,
		"current_docs": 0,
		"current_size_in_bytes": 0,
		"total": 26,
		"total_time_in_millis": 3576609,
		"total_docs": 408553,
		"total_size_in_bytes": 4984037156,
		"total_stopped_time_in_millis": 0,
		"total_throttled_time_in_millis": 215428,
		"total_auto_throttle_in_bytes": 102951098
	},
	"refresh": {
		"total": 329,
		"total_time_in_millis": 1571172,
		"external_total": 268,
		"external_total_time_in_millis": 650299,
		"listeners": 0
	},
	"flush": {
		"total": 45,
		"periodic": 40,
		"total_time_in_millis": 8580797
	},
	"warmer": {
		"current": 0,
		"total": 263,
		"total_time_in_millis": 1
	},
	"query_cache": {
		"memory_size_in_bytes": 0,
		"total_count": 10,
		"hit_count": 0,
		"miss_count": 10,
		"cache_size": 0,
		"cache_count": 0,
		"evictions": 0
	},
	"fielddata": {
		"memory_size_in_bytes": 0,
		"evictions": 0
	},
	"completion": {
		"size_in_bytes": 0
	},
	"segments": {
		"count": 132,
		"memory_in_bytes": 160112,
		"terms_memory_in_bytes": 67584,
		"stored_fields_memory_in_bytes": 82496,
		"term_vectors_memory_in_bytes": 0,
		"norms_memory_in_bytes": 0,
		"points_memory_in_bytes": 0,
		"doc_values_memory_in_bytes": 10032,
		"index_writer_memory_in_bytes": 0,
		"version_map_memory_in_bytes": 0,
		"fixed_bit_set_memory_in_bytes": 0,
		"max_unsafe_auto_id_timestamp": -1,
		"file_sizes": {}
	},
	"translog": {
		"operations": 0,
		"size_in_bytes": 275,
		"uncommitted_operations": 0,
		"uncommitted_size_in_bytes": 275,
		"earliest_last_modified_age": 77278667
	},
	"request_cache": {
		"memory_size_in_bytes": 0,
		"evictions": 0,
		"hit_count": 0,
		"miss_count": 0
	},
	"recovery": {
		"current_as_source": 0,
		"current_as_target": 0,
		"throttle_time_in_millis": 0
	}
},
```

**Comment by jmazanec15** [comment]
Hi @PavelArzhaev 

```
(1.1 * (4 * dimension + 8 * M), in my case dimension=512, M=16)
```
Calculation will estimate memory usage just for graph. For disk, there is a lot more written. We write the vectors as doc values as well.

In terms of other memory, how much data do you store in the other 2 fields?

**Comment by PavelArzhaev** [comment]
Hi @jmazanec15 
Thank you for the response!
Both graph and vector are written in documents (and so on disk), did I get it right?
In this case graph and vectors together would take ~2+2.4Gb per 1 million vectors, which is still 2x less than 10+Gb observed.
The only other fields stored are described in the first code section in the original post. I tried to remove knn field only and keep all other fields - the memory usage dropped significantly (less than 1Gb per 1 million records).

**Comment by jmazanec15** [comment]
@PavelArzhaev do you have replicas enabled?

> In this case graph and vectors together would take ~2+2.4Gb per 1 million vectors

So, vectors are also stored in the graph as well as doc values. This is one drawback of using nmslib and faiss. This redundancy is not present when using the lucene engine which was released with 2.2.

**Comment by PavelArzhaev** [comment]
@jmazanec15 replica shards are not allocated (so the index status is yellow).
I'm not sure about the plugin version, `/_cat/plugins` returns
```
{
	"name": "...",
	"component": "opensearch-knn",
	"version": "x.x.x.x"
},
```

**Comment by jmazanec15** [comment]
@PavelArzhaev What version of OpenSearch is the cluster running?

I will try to repro and figure out whats taking up more disk space.

**Comment by PavelArzhaev** [comment]
@jmazanec15 1.3, the latest available on AWS, if I'm not mistaken.

**Comment by jmazanec15** [comment]
@PavelArzhaev I see. The problem is that the vector is taking up a lot of space in _source in the stored fields. I created an index with 512 100 dimensional mappings and the files looked like this:
```
       405 _0.cfe
    985784 _0.cfs
       394 _0.si
    220356 _0_2011_vector_field.hnswc
       208 segments_2
         0 write.lock
```

The hnswc file takes up about the expected amount of disk. The majority of disk space is taken up by the cfs file. The cfs file is a compound file storing more files inside of it. I wrote a little [app](https://gist.github.com/jmazanec15/0a425812495dcac72526ec1a7b1543b3) to get the information of what files were in this file (actually a slightly different one than above because it was easier to use a different codec, but the file was about the same size) and it showed this:
```
_0.fnm: 692
_0_Lucene84_0.tip: 86
_0_Lucene84_0.doc: 78
_0_Lucene84_0.tim: 328
_0_Lucene80_0.dvd: 205817
_0.kdd: 377
_0.kdm: 143
_0_Lucene80_0.dvm: 407
_0.kdi: 68
_0_Lucene84_0.tmd: 186
_0.fdm: 158
_0.fdx: 103
_0.fdt: 777818
```
The dvd file is where the vectors are stored as doc values. The fdt file, taking up the most space, is the stored fields. One default stored field is the _source field. The _source field stores the json representation of the document you indexed. This will contain the vectors as well as the other fields. It is possible to disable the _source during indexing, but this comes with significant risks including the inability to reindex data.
